File: android-native/app/src/main/java/com/openflix/data/remote/dto/RemoteAccessDtos.kt
package com.openflix.data.remote.dto

import com.google.gson.annotations.SerializedName

/**
 * DTO for remote access status response.
 */
data class RemoteAccessStatusDto(
    @SerializedName("enabled") val enabled: Boolean,
    @SerializedName("connected") val connected: Boolean,
    @SerializedName("method") val method: String?,
    @SerializedName("tailscale_ip") val tailscaleIp: String?,
    @SerializedName("tailscale_hostname") val tailscaleHostname: String?,
    @SerializedName("magic_dns_name") val magicDnsName: String?,
    @SerializedName("backend_state") val backendState: String?,
    @SerializedName("login_url") val loginUrl: String?,
    @SerializedName("last_seen") val lastSeen: Long?,
    @SerializedName("error") val error: String?
)

/**
 * DTO for connection info response.
 */
data class ConnectionInfoDto(
    @SerializedName("server_url") val serverUrl: String,
    @SerializedName("network_type") val networkType: String,
    @SerializedName("is_remote") val isRemote: Boolean,
    @SerializedName("suggested_quality") val suggestedQuality: String,
    @SerializedName("tailscale_available") val tailscaleAvailable: Boolean?
)

/**
 * DTO for enable/disable response.
 */
data class RemoteAccessActionResponse(
    @SerializedName("success") val success: Boolean,
    @SerializedName("message") val message: String?,
    @SerializedName("login_url") val loginUrl: String?,
    @SerializedName("status") val status: RemoteAccessStatusDto?
)

/**
 * DTO for health check response.
 */
data class RemoteAccessHealthDto(
    @SerializedName("healthy") val healthy: Boolean,
    @SerializedName("checks") val checks: Map<String, Boolean>?,
    @SerializedName("warnings") val warnings: List<String>?
)

/**
 * DTO for install info response.
 */
data class TailscaleInstallInfoDto(
    @SerializedName("is_installed") val isInstalled: Boolean,
    @SerializedName("current_version") val currentVersion: String?,
    @SerializedName("install_command") val installCommand: String?,
    @SerializedName("configure_command") val configureCommand: String?,
    @SerializedName("doc_url") val docUrl: String?
)

/**
 * DTO for login URL response.
 */
data class TailscaleLoginUrlDto(
    @SerializedName("url") val url: String?,
    @SerializedName("expires_at") val expiresAt: Long?
)

// === Instant Switch DTOs ===

/**
 * DTO for instant switch status response.
 */
data class InstantSwitchStatusDto(
    @SerializedName("success") val success: Boolean,
    @SerializedName("data") val data: InstantSwitchDataDto?
)

data class InstantSwitchDataDto(
    @SerializedName("enabled") val enabled: Boolean,
    @SerializedName("active_channel") val activeChannel: String?,
    @SerializedName("cached_streams") val cachedStreams: Int,
    @SerializedName("total_memory_mb") val totalMemoryMB: Int,
    @SerializedName("recent_channels") val recentChannels: List<String>?
)

/**
 * Request to enable/disable instant switch.
 */
data class InstantSwitchEnabledRequest(
    @SerializedName("enabled") val enabled: Boolean
)

/**
 * Response for enable/disable instant switch.
 */
data class InstantSwitchEnabledResponse(
    @SerializedName("success") val success: Boolean,
    @SerializedName("enabled") val enabled: Boolean,
    @SerializedName("message") val message: String?
)

/**
 * DTO for cached streams response.
 */
data class CachedStreamsDto(
    @SerializedName("success") val success: Boolean,
    @SerializedName("cached") val cached: List<CachedStreamDto>?,
    @SerializedName("count") val count: Int
)

data class CachedStreamDto(
    @SerializedName("channel_id") val channelId: String,
    @SerializedName("buffered_bytes") val bufferedBytes: Int,
    @SerializedName("buffered_duration") val bufferedDuration: Double,
    @SerializedName("is_live") val isLive: Boolean,
    @SerializedName("last_access") val lastAccess: String?
)

/**
 * Request to set instant switch favorites.
 */
data class InstantSwitchFavoritesRequest(
    @SerializedName("favorites") val favorites: List<String>
)

// Type aliases for API compatibility
typealias RemoteAccessInstallInfoDto = TailscaleInstallInfoDto
typealias RemoteAccessLoginUrlDto = TailscaleLoginUrlDto

/**
 * DTO for remote streaming quality response.
 */
data class RemoteStreamingQualityDto(
    @SerializedName("quality") val quality: String,
    @SerializedName("available_qualities") val availableQualities: List<String>?
)

/**
 * Request to set remote streaming quality.
 */
data class SetStreamingQualityRequest(
    @SerializedName("quality") val quality: String
)

File: OpenFlix-tvOS/OpenFlix-tvOS/Models/DTOs/RemoteAccessDTOs.swift
import Foundation

// MARK: - Connection Info
struct ConnectionInfoDTO: Codable {
    let serverUrl: String
    let networkType: String
    let isRemote: Bool
    let suggestedQuality: String
    let tailscaleAvailable: Bool
}

// MARK: - Remote Access Status
struct RemoteAccessStatusDTO: Codable {
    let enabled: Bool
    let connected: Bool
    let method: String?
    let tailscaleIp: String?
    let tailscaleHostname: String?
    let magicDnsName: String?
    let backendState: String?
    let loginUrl: String?
    let lastSeen: Date?
    let error: String?
}

// MARK: - Remote Access Action Response
struct RemoteAccessActionResponse: Codable {
    let success: Bool
    let message: String?
    let loginUrl: String?
    let status: RemoteAccessStatusDTO?
}

// MARK: - Remote Access Health
struct RemoteAccessHealthDTO: Codable {
    let healthy: Bool
    let checks: [String: Bool]?
    let warnings: [String]?
}

// MARK: - Tailscale Install Info
struct TailscaleInstallInfoDTO: Codable {
    let isInstalled: Bool
    let currentVersion: String?
    let installCommand: String?
    let configureCommand: String?
    let docUrl: String?
}

// MARK: - Tailscale Login URL
struct TailscaleLoginUrlDTO: Codable {
    let url: String?
}

// MARK: - Network Type
enum NetworkType: String, Codable, CaseIterable {
    case wifi = "wifi"
    case cellular = "cellular"
    case vpn = "vpn"
    case ethernet = "ethernet"
    case unknown = "unknown"

    var displayName: String {
        switch self {
        case .wifi: return "Wi-Fi"
        case .cellular: return "Cellular"
        case .vpn: return "VPN"
        case .ethernet: return "Ethernet"
        case .unknown: return "Unknown"
        }
    }

    var icon: String {
        switch self {
        case .wifi: return "wifi"
        case .cellular: return "antenna.radiowaves.left.and.right"
        case .vpn: return "lock.shield"
        case .ethernet: return "cable.connector"
        case .unknown: return "questionmark.circle"
        }
    }
}

// MARK: - Instant Switch Status
struct InstantSwitchStatusDTO: Codable {
    let enabled: Bool
    let activeChannel: String?
    let cachedStreams: Int
    let totalMemoryMB: Int
    let predictions: [String]?
}

// MARK: - Cached Streams
struct CachedStreamsDTO: Codable {
    let streams: [CachedStreamDTO]
}

struct CachedStreamDTO: Codable {
    let channelId: String
    let bufferSize: Int
    let isLive: Bool
    let lastAccess: Date?
}

// MARK: - Remote Streaming Quality
enum RemoteStreamingQuality: String, Codable, CaseIterable {
    case original = "original"
    case quality1080p = "1080p"
    case quality720p = "720p"
    case quality480p = "480p"
    case quality360p = "360p"
    case auto = "auto"

    var displayName: String {
        switch self {
        case .original: return "Original"
        case .quality1080p: return "1080p"
        case .quality720p: return "720p"
        case .quality480p: return "480p"
        case .quality360p: return "360p"
        case .auto: return "Auto"
        }
    }

    var description: String {
        switch self {
        case .original: return "Best quality, highest bandwidth"
        case .quality1080p: return "Full HD, 8+ Mbps"
        case .quality720p: return "HD, 4+ Mbps"
        case .quality480p: return "SD, 2+ Mbps"
        case .quality360p: return "Low, 1+ Mbps"
        case .auto: return "Automatically adjust"
        }
    }

    var bandwidth: Int {
        switch self {
        case .original: return 0
        case .quality1080p: return 8000
        case .quality720p: return 4000
        case .quality480p: return 2000
        case .quality360p: return 1000
        case .auto: return -1
        }
    }

    static func suggestedFor(_ networkType: NetworkType) -> RemoteStreamingQuality {
        switch networkType {
        case .wifi, .ethernet:
            return .original
        case .vpn:
            return .quality720p
        case .cellular:
            return .quality480p
        case .unknown:
            return .auto
        }
    }
}
